feat: visualize runs on the canvas

doc-han opened the pull request:
## Description

This PR *[adds/changes/fixes]... (A description of your work goes here.)*

Closes #3387 

## Validation steps

1. *(How can a reviewer validate your work?)*

## Additional notes for the reviewer

1. *(Is there anything else the reviewer should know or look out for?)*

## AI Usage

Please disclose how you've used AI in this work (it's cool, we just want to know!):

- [x] Code generation (copilot but not intellisense)
- [x] Learning or fact checking
- [ ] Strategy / design
- [ ] Optimisation / refactoring
- [ ] Translation / spellchecking / doc gen
- [ ] Other
- [ ] I have not used AI

You can read more details in our [Responsible AI Policy](https://www.openfn.org/ai#pull-request-templates)

## Pre-submission checklist

- [x] I have performed a **self-review** of my code.
- [x] I have implemented and tested all related **authorization policies**. (e.g., `:owner`, `:admin`, `:editor`, `:viewer`)
- [x] I have updated the **changelog**.
- [x] I have ticked a box in "AI usage" in this PR

josephjclark commented:
Couple of early TODOs:
- Stub in some kind of tooltip on the pass/fail icon. This should show output for multiple executions of a single node
- A node is only a success if it never fails (across multiple executions)
- Nodes that were not executed need to be greyed out
- Disable edit mode while looking at a run (don't show the + icon on a node)

josephjclark commented:
Thoughts on showing multiple runs of the same node:
- render multiple edges?
- render the execution graph, not the workflow graph? Totally different tree?
- duplicate nodes in the workflow (is this the same as above?)
- show multiple status icons (3 ticks on cross)

For the moment, I want to do the simplest possible thing and show show as single status and tooltip.

doc-han commented:
AI chat icon already at the top-left corner of the screen. same place we want workflow history to be. 
<img width="897" height="623" alt="Screenshot 2025-07-15 at 9 49 28 PM" src="https://github.com/user-attachments/assets/e4a18eb9-f879-49fa-bae6-eb8b6aafb398" />

doc-han commented:
I have a situation where `starting_job_id` doesn't seem correct. strange

josephjclark commented:
Some small UX feedback notes from me:

- `Recent Activities` -> `Recent Activity`
- There are currently two collapse icons in the history panel - can we remove the "open in new window" one?
- The amount of repetition in this list bothers me. I know we're looking at Work Order and Runs here, but a) it's a lot of information, b) I wonder whether users understand this. Not convinced the UUIDs actually help anything either.
<img width="345" height="96" alt="image" src="https://github.com/user-attachments/assets/880444e6-b064-46eb-9cce-fb4bb780d678" />
- Generally, the UI indicators that say "You are currently looking at THIS historical run" are too subtle. It can be hard to know whether a run is loaded, and if so which. Especially if the history panel is collapsed. Maybe the Panel Title needs to be way more obvious? Maybe we need a banner above the diagram? Maybe we need more information alongside the snapshot id? I'm tempted to leave this until later, but I'm really worried that users landing on this page for the first time are going to have no idea what's happening
- We're currently showing the run duration on each step, in the label, but this looks horrible and upsets all the label layout. Can we put that in the tooltip or something?
- I think the "started from here" icon should be a star, not a play button, because it's not a verb and the play button looks like a verb
- As noted in a call recently, some alignments in the history panel look a little off

What do you think @taylordowns2000 ? I can make some of these changes directly. Maybe others we defer until later.

josephjclark commented:
- Update inspector close to not remove the run from the URL (Maybe)

doc-han commented:
"Switch to Latest" may be a toggle. 

reason: once you hit "Switch to Latest", it's difficult to get back to the snapshot for the run. You can do so now by deseleting the run and reselecting it.

doc-han commented:
### User Journey

#### 1. Wokflow Page to Mini-History
- Go to **Workflows Page**
- Select a **Workflow**
- Find the **Mini-History Panel** (top-left) **collapsed** because no run is selected

#### 2. Pick & Visualize a Run
- Click a run → it overlays on the workflow
- Click the run again to deselect and remove overlays

#### 3. Edit Enabled/Disabled
- If on **Latest Snapshot** → Editing is **enabled**
- If on an **Old Snapshot** → Editing is **disabled**

#### 4. Switch to Latest (Old Snapshots)
- Click **“Switch to Latest Version”** → Editing is now **enabled**
- Also, the runs(that happened on an old snapshot) are visualized on the latest snapshot

---
#### 1. History Page to Mini-History
- Go to **History Page**
- Select a **Run** from a Workorder and enter the canvas through there.
- Find the **Mini-History Panel** (top-left) **expanded** because a run is already selected.
- A run history should already be overlayed on the workflow.

Continue from **2.** in the First flow above...

josephjclark reviewed (changes requested):
Left a few small things which I think ought to be tidied up. Not every comment needs dealing with now - please just fix the small and easy stuff.

A wider note (for @taylordowns2000 really): most of my [UX comments](https://github.com/OpenFn/lightning/pull/3382#issuecomment-3113803963) are outstanding

Basically I am still concerned about the amount of confusion this UI will introduce to users. And sadly there's a lot of work that still needs to be done to get this feature really tight. I think we should aim to phase this feature in a bit more gently - but of course that will require more dev time.

doc-han commented on assets/js/workflow-diagram/MiniHistory.tsx:
```diff
@@ -0,0 +1,182 @@
+  };
+
+  const historyToggle = () => {
+    setIsCollapsed(p => !p);
```
I doubt it goes stale. since I'm getting it from the setAction itself.

elias-ba commented on lib/lightning/invocation.ex:
```diff
@@ -915,4 +915,25 @@ defmodule Lightning.Invocation do
     end
   end
+
+  def latest_workflow_runs_steps(workflow_id) do
+    WorkOrder
+    |> where([w], w.workflow_id == ^workflow_id)
+    |> order_by([w], desc: w.last_activity)
+    |> limit(1)
+    |> preload(
+      runs:
+        ^from(r in Lightning.Run,
+          order_by: [desc: r.started_at],
+          limit: 1,
+          preload: [:steps]
+        )
+    )
+    |> Repo.one()
+    |> case do
+      nil -> []
+      %{runs: [last_run]} -> last_run.steps
+      _ -> []
+    end
+  end
```
I love this function! I have a few suggestions that might help:

1. Could we consider a more descriptive name like `get_latest_run_steps_for_workflow`? The current name is a bit hard to parse and might be clearer if it indicates we're getting steps from the most recent run.

2. Can I suggest adding a brief doc comment to explain the logic? Something like:
```elixir
@doc """
Fetches steps from the most recent run of the latest work order for a given workflow.
Returns an empty list if no work order or runs exist.
"""
```

3. Would it be worth considering a slightly cleaner pattern match? Maybe something like:
```elixir
|> case do
  nil -> []
  %{runs: [run | _]} -> run.steps
  %{runs: []} -> []
end
```

These are just suggestions, the current implementation seems to work well! The use of preload with the subquery is actually quite elegant for getting exactly what we need in a single database round trip. 

What do you think?

doc-han replied:
This function is dead code. to be removed.

doc-han replied:
dead code removed.

elias-ba commented on lib/lightning/work_orders.ex:
```diff
@@ -484,6 +484,89 @@ defmodule Lightning.WorkOrders do
     {:ok, 0, 0}
   end
 
+  def get_run_steps(run_id) do
+    Run
+    |> where([r], r.id == ^run_id)
+    |> preload([:created_by, :steps])
+    |> Repo.one()
+    |> case do
+      nil ->
+        %{start_from: nil, steps: [], isTrigger: true}
+
+      %{steps: run_steps, starting_trigger_id: nil, starting_job_id: job_id} =
+          data ->
+        %{
+          start_from: job_id,
+          steps: run_steps,
+          isTrigger: false,
+          inserted_at: data.inserted_at,
+          run_by:
+            if(is_nil(data.created_by), do: nil, else: data.created_by.email)
+        }
+
+      %{steps: run_steps, starting_trigger_id: trigger_id, starting_job_id: nil} =
+          data ->
+        %{
+          start_from: trigger_id,
+          steps: run_steps,
+          isTrigger: true,
+          inserted_at: data.inserted_at
+        }
+
+      _ ->
+        %{start_from: nil, steps: [], isTrigger: true}
+    end
+  end
```
1. I noticed the function returns different keys depending on the case - sometimes we get `run_by` and `inserted_at`, sometimes not. Could we consider making the return structure consistent across all cases? This would make it more predictable for code consuming this function. Maybe something like:

```elixir
%{
  start_from: ...,
  steps: ...,
  isTrigger: ...,
  inserted_at: nil,  # always include, even if nil
  run_by: nil        # always include, even if nil
}
```

2. Would it be clearer if we simplified the case statement? Perhaps we could consolidate the logic like:

```elixir
case Repo.get(Run, run_id, preload: [:created_by, :steps]) do
  nil ->
    # return default structure
  run ->
    %{
      steps: run.steps,
      inserted_at: run.inserted_at,
      run_by: run.created_by && run.created_by.email,
      isTrigger: !is_nil(run.starting_trigger_id),
      start_from: run.starting_job_id || run.starting_trigger_id
    }
end
```

3. Can I ask about the `_ ->` pattern? It seems like this would only match if a run has both `starting_trigger_id` and `starting_job_id` set. I don't think this a valid state in our data model

4. Would it help to add a `@spec` to document what this function returns? It could help future developers understand the expected output structure.

Just some thoughts! The logic for determining trigger vs job runs is clear, which is great. What do you think about these suggestions?

midigofrank replied:
@doc-han can we move the mapping logic over to the Liveview? This file contains logic that can be reused/called out by other modules. 

Actually, I don't think we need this function here at all. I would recommend that you do this in the liveview.
You can get the run by calling `Lightning.Runs.get(run_id, include: [:created_by, :steps]` and then doing the mapping logic in the liveview

elias-ba commented on lib/lightning/work_orders.ex:
```diff
@@ -484,6 +484,89 @@ defmodule Lightning.WorkOrders do
+      )
+
+    workorder_with_run =
+      if run_id do
+        from([wo, r] in main_query,
+          where: r.id != ^run_id
+        )
+      else
+        main_query
+      end
+
+    final_query =
+      if run_id do
+        from(wo in WorkOrder,
+          join: r in assoc(wo, :runs),
+          where: r.id == ^run_id and wo.workflow_id == ^workflow_id,
+          preload: [:snapshot, runs: [:steps]],
+          union: ^workorder_with_run
+        )
+      else
+        workorder_with_run
+      end
+
+    final_query
+    |> order_by(desc: fragment("last_activity"))
+    |> Repo.all()
+    |> Enum.map(fn worder ->
+      %{
+        runs:
+          worder.runs
+          |> Enum.map(fn run ->
+            Map.take(run, [:id, :state, :error_type, :started_at, :finished_at])
+          end),
+        version: worder.snapshot.lock_version,
+        state: worder.state,
+        last_activity: worder.last_activity,
+        id: worder.id
+      }
+    end)
+  end
```
Ah ! I remember this big boy 😆. It is handling some complex logic, and I have a few thoughts that might help make it clearer and more maintainable:

1. I'm finding the UNION logic a bit hard to follow. Could we add a comment explaining the business requirement? It seems like when `run_id` is provided, we want to ensure that specific work order is included plus up to 19 others. Is that correct? Would it be possible to simplify this approach?

2. I noticed we're loading all runs and all steps for each work order. Do we need all this data upfront? If work orders have many runs, this could become a performance bottleneck. Could we consider:
- Loading only recent runs (maybe last 5-10)?
- Or lazy-loading runs when needed?
- Adding a comment about why we need all runs if it's a business requirement?

3. Could we replace `fragment("last_activity")` with `[wo], desc: wo.last_activity`? Using the Ecto syntax would be safer and more maintainable.

4. What do you think about extracting the formatting into helper functions? Something like:

```elixir
|> Enum.map(&format_workorder_summary/1)

defp format_workorder_summary(workorder) do
  %{
    id: workorder.id,
    state: workorder.state,
    last_activity: workorder.last_activity,
    version: workorder.snapshot.lock_version,
    runs: Enum.map(workorder.runs, &format_run_summary/1)
  }
end

defp format_run_summary(run) do
  Map.take(run, [:id, :state, :error_type, :started_at, :finished_at])
end
```

5. Have you considered if we could handle the "highlight specific run" logic differently? Maybe we could:
1. Always fetch the 20 most recent work orders
2. If a `run_id` is provided, check if its work order is in the list
3. If not, fetch it separately and prepend it to the results

Just thinking this might be simpler than using UNION?

These are just suggestions to consider! The current implementation does work, and this business logic here is not quite easy. What are your thoughts on any of these points?

elias-ba commented on lib/lightning_web/live/workflow_live/edit.ex:
```diff
@@ -517,9 +517,9 @@ defmodule LightningWeb.WorkflowLive.Edit do
-                            (!is_nil(@workflow.deleted_at) or !@can_edit_workflow or
-                               @has_child_edges or @is_first_job or
-                               @snapshot_version_tag != "latest") ||
+                            !is_nil(@workflow.deleted_at) or !@can_edit_workflow or
+                              @has_child_edges or @is_first_job or
+                              @snapshot_version_tag != "latest" ||
```
I think it's time we start extracting these disabled logic into functions for maintainability. But that is out of scope of this work. Just wondering if we should have an issue to track it.

elias-ba commented on lib/lightning_web/live/workflow_live/edit.ex:
```diff
@@ -1520,7 +1522,26 @@ defmodule LightningWeb.WorkflowLive.Edit do
+      )
+
+    # don't forget to send update state of disabled
+    socket =
+      socket
+      |> maybe_disable_canvas()
+
+    {:reply,
+     %{
+       workflow_params: socket.assigns.workflow_params,
+       run_steps: run_steps,
+       run_id: run_id,
+       history: history
+     }, socket}
```
We could do this, it's more elixir than updating the socket

```elixir 
    {:reply,
     %{
       workflow_params: socket.assigns.workflow_params,
       run_steps: run_steps,
       run_id: run_id,
       history: history
     }, maybe_disable_canvas(socket)}
  ```

elias-ba commented on lib/lightning_web/live/workflow_live/edit.ex:
```diff
@@ -2168,6 +2189,30 @@ defmodule LightningWeb.WorkflowLive.Edit do
     {:noreply, socket}
   end
 
+  defp get_run_steps_and_history(workflow_id, run_id) do
+    run_steps =
+      if run_id == nil do
+        %{steps: []}
+      else
+        WorkOrders.get_run_steps(run_id)
+      end
+      |> Map.update!(:steps, fn steps ->
+        Enum.map(steps, fn step ->
+          %{
+            job_id: step.job_id,
+            error_type: step.error_type,
+            exit_reason: step.exit_reason,
+            started_at: step.started_at,
+            finished_at: step.finished_at
+          }
+        end)
+      end)
+
+    history = WorkOrders.get_workorders_with_runs(workflow_id, run_id)
+
+    %{run_steps: run_steps, history: history}
+  end
```
1. I'm a bit concerned about the `Map.update!` call when `run_id` is nil. Since we return `%{steps: []}` in that case, the `Map.update!` will work, but we might be missing other expected keys. If `WorkOrders.get_run_steps/1` returns additional fields (like `:isTrigger`, `:start_from`, etc.), could this cause issues elsewhere?  Would something like this be safer?

```elixir
run_steps = 
  if run_id do
    WorkOrders.get_run_steps(run_id)
    |> Map.update!(:steps, &format_steps/1)
  else
    %{steps: [], isTrigger: nil, start_from: nil} # include all expected keys
  end
```

2. Could we extract the step transformation into a helper function for better readability?

```elixir
defp format_step(step) do
  %{
    job_id: step.job_id,
    error_type: step.error_type,
    exit_reason: step.exit_reason,
    started_at: step.started_at,
    finished_at: step.finished_at
  }
end

# Then use it like:
|> Map.update!(:steps, fn steps -> Enum.map(steps, &format_step/1) end)
```

3. I notice we're fetching run data twice - once in `get_run_steps` and potentially again in `get_workorders_with_runs`. Is there an opportunity to optimize this, or is it necessary for the business logic?

4. Just a thought - since this is a private function, the name is fine, but would `get_run_details_and_workflow_history` be more descriptive of what it actually does?

I love the structure and the separation of concerns between run steps and history. What do you think about these suggestions?

doc-han replied:
1. Yeah we can add all the expected keys. [**done**]
2. Yeah that's possible. I don't see this being reused anywhere hence the reason I left it there but yes can be extracted. [**done**]
3. They're two difference cases. `get_run_steps` gets the steps of a run whilst `get_workorder_with_runs` gets several workorders and their corresponding runs.
- `get_workorder_with_runs` is for getting the workorder and runs to be shows as the history of the workflow
- `get_run_steps` gets the steps involved in a run when a user select a run to view.

4. `get_workorders_with_runs` (getting workorders with their runs) vs `get_run_details_and_workflow_history`. I'll for the first one. what do you say?

elias-ba commented on lib/lightning_web/live/workflow_live/edit.ex:
```diff
@@ -2802,13 +2851,41 @@ defmodule LightningWeb.WorkflowLive.Edit do
+
+    # snapshots currently don't have positions
+    # if we don't do this, the patch wouldn't replace the existing position in params
+    updated_snapshot = Map.put(snapshot, :positions, nil)
```
Should we add `positions` as a field to the Snapshot schema instead of doing this manual `Map.put`? Doing this manual manipulation feels fragile and could break if the Snapshot structure changes. Having `positions` as an actual field with a default of `nil` would make the data structure more explicit and eliminate the need for this workaround.

Something like:

```elixir
schema "workflow_snapshots" do
  # ... existing fields ...
  field :positions, :map, default: nil
  # ... rest of schema ...
end
```

What do you think?

doc-han replied:
This is cool because we'll obviously have to support node positions in snapshots too. 
with this I run a migration after right?

elias-ba reviewed (changes requested):
Hey @doc-han, this is really huge work! I loved reviewing it, nice work 👏🏽.

I left a few suggestions/questions throughout, nothing really blocking, as this seems to work well. Some of my questions are more related to maintainability, clarity, and performance, but they may be out of scope for this PR. In those cases, we can just open issues to track them separately rather than blocking this work.

Great job overall! Let me know if you'd like to discuss any of the points I raised.

midigofrank commented on lib/lightning/work_orders.ex:
```diff
@@ -484,6 +484,89 @@ defmodule Lightning.WorkOrders do
+    final_query
+    |> order_by(desc: fragment("last_activity"))
+    |> Repo.all()
+    |> Enum.map(fn worder ->
```
Can we move the mapping logic to the liveview?

midigofrank commented on lib/lightning_web/live/workflow_live/edit.ex:
```diff
@@ -2765,6 +2810,10 @@ defmodule LightningWeb.WorkflowLive.Edit do
         handle_method_assignment(socket, method)
 
+      %{"m" => "history", "a" => run_id, "v" => version_tag, "s" => selected_id}
+      when not is_nil(version_tag) ->
```
```suggestion
      when is_binary(version_tag) ->
```

taylordowns2000 reviewed (dismissed):
Looks great @doc-han ! It feels so stable and simple. (And I know it's _anything but simple_, so you should feel very proud that it now _feels_ simple.)

A few things:
1. please rebase on/merge `main` and resolve conflicts
2. the link to "go to the history page, filtered by this workflow" doesn't work and is not _quite_ vertically aligned (seems low compared to the text) as far as i can tell (image 1)
3. when there aren't many runs for a workflow, the table is still too large. it should have a max height, but not a min. (image 2)
4. when there is no activity, it should say "No related activity" rather than "No recent activity" - because "recent" makes it sound like there might be older activity. (image 3)

-------- new issues to create --------

5. please use the `.datetime` "relative" components (or use the same hook in react) for datetimes in the mini history table so we keep the standardization now on `main`?
6. which run does it show when there are 2? highest exit code? (can we show some sort of _count_ of how many runs there are in that step? image 4)
7. indicate when the loaded run doesn't match the nodes on `latest` and can't be fully visualized
8. use standard state `pills` for list, use standard step icons as `step_icons` on canvas for step exit codes
9. when a run is loaded, we should have the option to retry or create a new work order, exactly like the button on the inspector (image 5)

## image 1
<img width="153" height="86" alt="Image" src="https://github.com/user-attachments/assets/5dc2dae5-82d4-421b-8f95-dbd740d615f7" />

## image 2

<img width="428" height="430" alt="Image" src="https://github.com/user-attachments/assets/03e52f60-559a-4724-ac08-0a193a4d35e6" />

## image 3

<img width="616" height="591" alt="Image" src="https://github.com/user-attachments/assets/3d0df29d-b836-402b-ae93-4c881725fddd" />

## image 4

<img width="408" height="405" alt="Image" src="https://github.com/user-attachments/assets/2f1f1132-7d7f-405e-9b7c-ca90e68076c1" />

## image 5

<img width="325" height="124" alt="Image" src="https://github.com/user-attachments/assets/25b012d4-5b50-452e-806c-b65aed90aa31" />

doc-han commented on assets/js/workflow-diagram/MiniHistory.tsx:
```diff
@@ -0,0 +1,182 @@
+  }, [selectedItem]);
+
+  const expandWorkorderHandler = (workorder: WorkflowRunHistory[number]) => {
+    if (workorder.runs.length === 1 && workorder.runs[0]) {
```
ts compiler isn't smart enough to know that when length is 1 then index 0 should be accessible.

doc-han commented on lib/lightning/work_orders.ex:
```diff
@@ -484,6 +484,89 @@ defmodule Lightning.WorkOrders do
+      end
+
+    final_query
+    |> order_by(desc: fragment("last_activity"))
```
tested. didn't seem to work. 

I've simplified the query. 
1. get recent 20 workorders
2. get workorder that contains the `run_id` when `run_id` isn't `nil`.
3. merge them with a union.

josephjclark commented on assets/js/workflow-diagram/nodes/Node.tsx:
```diff
@@ -83,15 +97,32 @@ const Node = ({
   errors,
+  type,
 }: BaseNodeProps) => {
+  const isTriggerNode = type === 'trigger';
```
I don't think we need to do this now: a comment in the code reminding us to refactor later is fine :) More important to get the work merged

doc-han commented:
> Looks great @doc-han ! It feels so stable and simple. (And I know it's _anything but simple_, so you should feel very proud that it now _feels_ simple.)
> 
> A few things:
> 
> 1. please rebase on/merge `main` and resolve conflicts
> 2. the link to "go to the history page, filtered by this workflow" doesn't work and is not _quite_ vertically aligned (seems low compared to the text) as far as i can tell (image 1)
> 3. when there aren't many runs for a workflow, the table is still too large. it should have a max height, but not a min. (image 2)
> 4. when there is no activity, it should say "No related activity" rather than "No recent activity" - because "recent" makes it sound like there might be older activity. (image 3)
> 5. please use the `.datetime` "relative" components (or use the same hook in react) for datetimes in the mini history table so we keep the standardization now on `main`?
> 


1. synced with main [**resolved**]
2. go to history vertically aligned & implemented to work [**resolved**]
3. height of mini-history made to fit [**resolved**]
4. wording updated to "No related activity. [**resolved**]
5. Not clear on that yet

doc-han commented:
Issue created for updating datetime format on the history panel. 
#3499

elias-ba commented:
Hey @doc-han, my commit here does two things:

### 1. Add positions to snapshot creation

I added `positions` to the cast and build functions of the snapshot data model, this allows workflow positions to be  preserved when creating snapshots

### 2. Fix flaky ordering in `get_workorders_with_runs/2`

While adding tests for the workorders query functions, I discovered the test `"orders workorders by last_activity descending"` was flaky - it would pass/fail randomly.

**The issue:** The original function used `order_by(desc: fragment("last_activity"))` which caused non-deterministic ordering due to:
- Ambiguous column references when joining `work_orders` with `runs`
- Conflicts between `DISTINCT` and `ORDER BY` clauses
- Complex UNION queries when `run_id` is provided

**The fix:** I rewrote the function to:
- Eliminate UNION queries and fragments
- Use explicit column references
- Handle "with run_id" and "without run_id" cases separately
- Ensure deterministic ordering

This test proves the fix - it fails with the original implementation but passes consistently with the rewrite.

https://github.com/OpenFn/lightning/blob/6b9aec3770dfae533dff8c3bee3f4cd052f3279d/test/lightning/work_orders_test.exs#L2766

**Please verify:** The UI still works correctly with these changes. I did some manual testing and everything looked good, but would appreciate if you could double-check the workorders display still behaves as expected.

Feel free to review [[the commit](https://github.com/OpenFn/lightning/pull/3382/commits/6b9aec3770dfae533dff8c3bee3f4cd052f3279d)](https://github.com/OpenFn/lightning/pull/3382/commits/6b9aec3770dfae533dff8c3bee3f4cd052f3279d) and let me know if you have any concerns about the approach.

elias-ba commented:
In this commit ([6d0e031](https://github.com/OpenFn/lightning/pull/3382/commits/6d0e031f60196f62b38d60a1509f68758b47c935)), I am adding few liveview tests to increase coverage in the Workflow Edit module.

elias-ba reviewed (approved):
Great work @doc-han, this feature is a banger

elias-ba commented:
You need to add a Changelog line before merging this

midigofrank reviewed (changes requested):
Nicely done @doc-han , this now look ready to go. I just have one last request, can we get rid of `WorkOrders.get_run_steps/1`:
```elixir

  def get_run_steps(run_id) do
    Run
    |> where([r], r.id == ^run_id)
    |> preload([:created_by, :steps])
    |> Repo.one()
  end
 ```
 
Given that this function fetches a run, we can utilize the `Runs.get/2`:
```elixir
Runs.get(run_id, include: [:created_by, :steps])
```

doc-han commented:
> Nicely done @doc-han , this now look ready to go. I just have one last request, can we get rid of `WorkOrders.get_run_steps/1`:
> 
> ```elixir
>   def get_run_steps(run_id) do
>     Run
>     |> where([r], r.id == ^run_id)
>     |> preload([:created_by, :steps])
>     |> Repo.one()
>   end
> ```
> 
> Given that this function fetches a run, we can utilize the `Runs.get/2`:
> 
> ```elixir
> Runs.get(run_id, include: [:created_by, :steps])
> ```

Resolved. I've removed the function and used `Runs.get` directly as suggested. 
[ba14ab0](https://github.com/OpenFn/lightning/pull/3382/commits/ba14ab0c602028b737099daddcc53240aceb3c34)
Thanks man.

midigofrank reviewed (approved):
Nicely done

doc-han commented:
> Hey @doc-han, my commit ([6b9aec3](https://github.com/OpenFn/lightning/pull/3382/commits/6b9aec3770dfae533dff8c3bee3f4cd052f3279d) here does two things:
> 
> ### 1. Add positions to snapshot creation
> I added `positions` to the cast and build functions of the snapshot data model, this allows workflow positions to be preserved when creating snapshots
> 
> ### 2. Fix flaky ordering in `get_workorders_with_runs/2`
> While adding tests for the workorders query functions, I discovered the test `"orders workorders by last_activity descending"` was flaky - it would pass/fail randomly.
> 
> **The issue:** The original function used `order_by(desc: fragment("last_activity"))` which caused non-deterministic ordering due to:
> 
> * Ambiguous column references when joining `work_orders` with `runs`
> * Conflicts between `DISTINCT` and `ORDER BY` clauses
> * Complex UNION queries when `run_id` is provided
> 
> **The fix:** I rewrote the function to:
> 
> * Eliminate UNION queries and fragments
> * Use explicit column references
> * Handle "with run_id" and "without run_id" cases separately
> * Ensure deterministic ordering
> 
> This test proves the fix - it fails with the original implementation but passes consistently with the rewrite.
> 
> https://github.com/OpenFn/lightning/blob/6b9aec3770dfae533dff8c3bee3f4cd052f3279d/test/lightning/work_orders_test.exs#L2766
> 
> **Please verify:** The UI still works correctly with these changes. I did some manual testing and everything looked good, but would appreciate if you could double-check the workorders display still behaves as expected.
> 
> Feel free to review the code in the commit and let me know if you have any concerns about the approach.

I've tested several times and everything seems to work just fine. Thanks for this @elias-ba 
really appreciated. ❤️

doc-han commented:
This PR actually closes #3387 which is a sub-issue of #3103
